DCP-1600 init

jimmymadrigal opened the pull request:
Initial Validation MVP for DSCVR frames payloads.
The idea is to make this library as simple and lightweight as posible.

Here are some highlights:
- Provides full validation connecting to the Dscvr GraphQL api (to staging for the moment)
- GraphQL api could be an optional function param (otherwise defaults to staging)
- There is one TODO for timestamps which should be removed when the Proxy is finished
- Uses urql to query Graphql and nothing else (no typescript plugins nor type generators)
- @open-frames/types MUST BE a devDependency, otherwise frames.js won't compile

<img width="926" alt="image" src="https://github.com/dscvr-one/frames-adapter/assets/11619520/4e41f01e-9af3-4f3f-8a34-5e14cd11f724">

jimmymadrigal committed:
feat: validation MVP

jimmymadrigal committed:
fix: consider timestamp as string scneario

jimmymadrigal committed:
fix: contentId comes as a string in the payload

vinayakkulkarni commented on .github/workflows/release-package.yml:
```diff
@@ -0,0 +1,38 @@
+    types: [created]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          node-version: '20.x'
+      - run: npm ci
```
```suggestion
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repository 🎉
        uses: actions/checkout@v4

      - name: Setup node env 📦
        uses: actions/setup-node@v4
        with:
          node-version-file: .nvmrc
          check-latest: true
          cache: 'npm'

      - name: Install dependencies 🚀
        run: npm ci --prefer-offline --no-audit
```

jimmymadrigal replied:
Fixed

vinayakkulkarni commented on .github/workflows/release-package.yml:
```diff
@@ -0,0 +1,38 @@
+          node-version: '20.x'
+      - run: npm ci
+
+  publish:
+    needs: build
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          node-version: '20.x'
+          registry-url: https://npm.pkg.github.com/
+      - run: npm install
+      - run: npm run build
+      - run: npm pack
+      - run: npm publish --access=public *.tgz
+        env:
+          NODE_AUTH_TOKEN: ${{secrets.GITHUB_TOKEN}}
+      - uses: actions/upload-artifact@v4
+        with:
+          name: release-package
+          path: ./*.tgz
```
```suggestion
  publish:
    needs: build
    runs-on: ubuntu-latest
    permissions:
      contents: read
      packages: write
    steps:
      - name: Check out repository ✨
        uses: actions/checkout@v4

      - name: Setup node env 📦
        uses: actions/setup-node@v4
        with:
          registry-url: https://npm.pkg.github.com/
          node-version-file: .nvmrc
          check-latest: true
          cache: 'npm'

      - name: Install dependencies 🚀
        run: npm ci --prefer-offline --no-audit

      - name: Run build 🏁
        run: npm run build

      - name: Create compressed tarball file 📦
        run: npm pack

      - name: Publish to GitHub Packages 🚀
        run: npm publish --access=public *.tgz
        env:
          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - name: Upload release package 📦
        uses: actions/upload-artifact@v4
        with:
          name: release-package
          path: ./*.tgz
```

jimmymadrigal replied:
Fixed

vinayakkulkarni commented on .gitignore:
Kindly update the `.gitignore` from the default template provided by GitHub for all node based repositories: https://github.com/github/gitignore/blob/main/Node.gitignore

jimmymadrigal replied:
Fixed

vinayakkulkarni commented on .nvmrc:
```diff
@@ -0,0 +1 @@
+20
```
```suggestion
20

```

jimmymadrigal replied:
Fixed

vinayakkulkarni commented on src/index.ts:
```diff
@@ -0,0 +1,2 @@
+export * from './types';
+export * from './validation';
```
can we avoid barrel files?

jimmymadrigal replied:
Fixed

vinayakkulkarni commented on src/default.ts:
```diff
@@ -0,0 +1,2 @@
+export const DEFAULT_DSCVR_API_URL =
+  'https://edge1-analytics.stg.dscvr.cloud/graphql';
```
should this be publicly exposed?

ncpenke replied:
good question - @jimmymadrigal this should be `https://dev.dscvr.one/graphql`. It's not setup yet, but for commit purposes let's please use that.

jimmymadrigal replied:
Fixed

vinayakkulkarni commented on .github/workflows/release-package.yml:
```diff
@@ -0,0 +1,38 @@
+name: 'Testing Release Package'
```
```suggestion
name: 'Build and release'
```

jimmymadrigal replied:
Fixed

vinayakkulkarni commented:
I also suggest adding [ship.js](https://github.com/algolia/shipjs) or [release-it](https://github.com/release-it/release-it) for automated releases instead of manually running the workflow file. 

This is also missing commitlint / husky which is standard practice across our repos? we want to add it?

Thanks.

ncpenke commented on src/validation.ts:
```diff
@@ -0,0 +1,137 @@
+  untrusted: ValidationType
+): boolean => {
+  if (!validated && !untrusted) {
+    return true;
```
Why is it `true` in this case? Feels like this method should simply be ```return validated && untrusted && validated === untrusted```

ncpenke replied:
Oh I get it, if it's not present in the untrusted or validated then it's also ok. I would simply this to:

```
return (!validated && !untrusted) || (validated === untrusted)
```

jimmymadrigal replied:
Fixed

ncpenke commented on src/validation.ts:
```diff
@@ -0,0 +1,137 @@
+  validatedResult: DscvrValidationResponse,
+  untrustedData: DscvrUntrustedData
+): boolean => {
+  if (
```
No need for the `if` here can simply be `return <conditional expression>`

jimmymadrigal replied:
Fixed

ncpenke commented on src/validation.ts:
```diff
@@ -0,0 +1,137 @@
+  return clientProtocol.startsWith(dscvrClientProtocolPrefix);
+};
+
+const validateUntrustedDataValue = (
```
I think this can be `validateField`

jimmymadrigal replied:
Fixed

ncpenke commented on src/validation.ts:
```diff
@@ -0,0 +1,137 @@
+  return false;
+};
+
+const validateUntrustedData = (
```
I think this can be `validateData`

jimmymadrigal replied:
Fixed

ncpenke commented on src/types.ts:
```diff
@@ -0,0 +1,29 @@
+}
+
+export type DscvrValidationResponse = {
+  validatedDscvrId: string;
```
actually I don't think we need the `validated` prefix in front of these

jimmymadrigal replied:
Fixed

ncpenke commented on src/validation.ts:
```diff
@@ -0,0 +1,137 @@
+      validatedResult.validatedDscvrId,
+      untrustedData.dscvrId
+    ) &&
+    validateUntrustedDataValue(
```
If you follow my suggestion above, then you can just loop through the superset of the keys in the trusted and untrusted data

jimmymadrigal replied:
Hmm, thats tricky, TS does not have a way to loop through the superset of a type, only for an object, and optional props wont appear in that loop if undefined.

The way to do this with a loop is this, but I honestly prefer the explicit if, WDYT?

```
  const keys: (keyof DscvrValidationResponse)[] = 
  [
    'dscvrId',
    'contentId',
    'buttonIndex',
    'url',
    'inputText',
    'state',
    'timestamp',
  ];
  keys.reduce((valid, key) => {
    return valid && validateField(validatedResult[key], untrustedDataChanged[key]);
  }, true);
```

ncpenke replied:
The only downside to this is the untrusted data can have fields that are not in the validated response and we wouldn't catch them.

Not to create more work, but how about setting any fields that are null in `validateField` to undefined, and then you can make sure the number of keys in untrusted is the same as the valid ones and also loop through all the keys and make sure they match up. Wdyt?

jimmymadrigal replied:
Omg, you are absolutely right, I didn't consider the fact that untrusted data could have extra attributes
thanks I changed the code

ncpenke commented on src/validation.ts:
```diff
@@ -0,0 +1,137 @@
+    validateUntrustedDataValue(validatedResult.state, untrustedData.state) &&
+    validateUntrustedDataValue(
+      validatedResult.timestamp,
+      Number(untrustedData.timestamp) // TODO: remove this when Proxy is fixed
```
not needed, since you already convert to `Number` in `validateFramesPost`

jimmymadrigal replied:
No but this is the untrusted, **validateFramesPost**  converts to number the`validatedResult`

ncpenke commented:
Looks really good! Just some feedback for clean up

ncpenke commented on README.md:
```diff
@@ -1 +1,98 @@
 # Frames Adapter
+
+Frames Adapter is a small typescript library designed to validate the payload from a signed frames POST call for DSCVR frames.
```
frames-adapter is a small typescript library that can be used as an adapter to different frames frameworks. Currently, frames.js is supported. The library provides a couple of helper methods to validate the trusted message data of a frame action and also to determine if a frame is using the DSCVR protocol.

jimmymadrigal replied:
Fixed

ncpenke commented on README.md:
```diff
@@ -1 +1,98 @@
 # Frames Adapter
```
I believe this should be `frames-adapter` (to mirror the package name)

jimmymadrigal replied:
Fixed

jimmymadrigal committed:
fix: pr feedback

Signed-off-by: Jimmy Madrigal <jimmymadrigal+github@gmail.com>

jimmymadrigal commented:
> I also suggest adding [ship.js](https://github.com/algolia/shipjs) or [release-it](https://github.com/release-it/release-it) for automated releases instead of manually running the workflow file.
> 
> This is also missing commitlint / husky which is standard practice across our repos? we want to add it?
> 
> Thanks.

Ok I created the ship.config.cjs file, removed the manual workflow and added the shipjs-trigger workflow. Hopefully this will create the public package to install, when the branch is merged

ncpenke commented on README.md:
```diff
@@ -1 +1,104 @@
-# Frames Adapter
+# frames-adapter
+
+frames-adapter is a small typescript library that can be used as an adapter to different frames frameworks. Currently, frames.js is supported. The library provides a couple of helper methods to validate the trusted message data of a frame action and also to determine if a frame is using the DSCVR protocol.
```
Sorry one more edit here:

```
frames-adapter is a small typescript library that can be used as an adapter to DSCVR's protocol from different frames frameworks. Currently, only frames.js is supported. 

The library provides a couple of helper methods to validate the trusted message data of a frame action and also to determine if a frame is using the DSCVR protocol.
```

ncpenke commented:
This is looking really good - just couple more comments from me.

ncpenke commented:
@jimmymadrigal can we remove package-json.lock here since it's a library?

jimmymadrigal commented:
> @jimmymadrigal can we remove package-json.lock here since it's a library?

@ncpenke according to this article you are right. Its weird that https://github.com/github/gitignore/blob/main/Node.gitignore does not provide that on their example.

yet if you go to a library like https://github.com/lodash/lodash it does not include it.

So I removed it, thanks!

ncpenke reviewed (approved):
Looks great!

vinayakkulkarni commented on src/validation.ts:
```diff
@@ -0,0 +1,123 @@
+  buttonIndex: number;
+}
+
+const validationQuery = `#graphql
+    query validate($message: String!) {
+      unpackFrameMessage(message: $message) {
+        user {
+            id
+        },
+        content {
+            id
+        }
+        buttonIndex,
+        state,
+        url,
+        timestamp,
+        inputText,
+      }
+    }
+`;
```
suggestion would be to move this piece of code to `src/queries/validation.query.ts` ?

vinayakkulkarni replied:
not a blocker

jimmymadrigal replied:
Done

vinayakkulkarni commented on src/validation.ts:
```diff
@@ -0,0 +1,123 @@
+  payload: DscvrFramesRequest,
+  apiUrl = DEFAULT_DSCVR_API_URL
+): Promise<DscvrValidationResponse> => {
+  const client = new Client({
+    url: apiUrl,
+    exchanges: [cacheExchange, fetchExchange],
+  });
```
maybe we create a Client in `src/clients/index.ts` and use it like

```js
const { client } = useUrqlClient({ url: apiUrl }); 
```

or even add `apiUrl` in `clients/index.ts` and just do `const { client } = useUrqlClient();`

again, not a blocker, just a suggestion.

jimmymadrigal replied:
Done

vinayakkulkarni commented on .husky/pre-push:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npm exec --no -- lint-staged --no-stash
```
seems there's a missing `lint-staged.config.cjs` ?

jimmymadrigal replied:
Thanks forgot about eslint and prettier, Done

vinayakkulkarni reviewed (approved):
Great work @jimmymadrigal! LGTM, some minor suggestions, not blockers.

jimmymadrigal committed:
fix: esling and prettier

Signed-off-by: Jimmy Madrigal <jimmymadrigal+github@gmail.com>

vinayakkulkarni commented on src/validation.ts:
```diff
@@ -0,0 +1,103 @@
+import { useUrqlClient } from './clients';
+import { validationQuery } from './queries/validation.query';
+
+type ValidationType = string | number | bigint | null | undefined;
+interface ValidatedQueryResult {
+  user: {
+    id: string;
+  };
+  content?: {
+    id: string;
+  };
+  state: string | null;
+  inputText: string | null;
+  timestamp: string;
+  url: string;
+  buttonIndex: number;
+}
```
maybe we could move these to `./types` and import from there to make this code more cleaner?

jimmymadrigal replied:
Done

vinayakkulkarni commented on src/types.ts:
```diff
@@ -0,0 +1,21 @@
+  contentId?: string;
+  state?: string;
+};
+export const dscvrClientProtocolPrefix = 'dscvr@' as const;
```
shouldn't this be `src/constants.ts` as `src/types.ts` should only export types & should be used like:

`import type { <type> } from './types';`

jimmymadrigal replied:
Done

vinayakkulkarni commented on src/index.ts:
```diff
@@ -0,0 +1,8 @@
+  type DscvrFramesRequest,
+  type DscvrUntrustedData,
+  type DscvrValidationResponse,
+  dscvrClientProtocolPrefix,
```
I suggest moving `dscvrClientProtocolPrefix` to `src/constants.ts` instead of `src/types.ts` as it's an `const` that is being exported rather than a `type`

jimmymadrigal replied:
Done

vinayakkulkarni commented on src/validation.ts:
```diff
@@ -0,0 +1,103 @@
+import {
+  type DscvrFramesRequest,
+  type DscvrValidationResponse,
+  dscvrClientProtocolPrefix,
```
I suggest moving `dscvrClientProtocolPrefix` to `src/constants.ts` instead of `src/types.ts` as it's an `const` that is being exported rather than a `type`

jimmymadrigal replied:
Done

vinayakkulkarni commented on src/validation.ts:
```diff
@@ -0,0 +1,103 @@
+  type DscvrFramesRequest,
+  type DscvrValidationResponse,
+  dscvrClientProtocolPrefix,
+  DscvrUntrustedData,
```
```suggestion
  type DscvrUntrustedData,
```

jimmymadrigal replied:
Done

jimmymadrigal committed:
fix: more pr feedback

Signed-off-by: Jimmy Madrigal <jimmymadrigal+github@gmail.com>
